### ISSUE TITLE
Database configuration needed even when it's not used

### ISSUE THREAD

**Issue by josecelano** [torrust/torrust-tracker#999]
Usually, when you run the tracker for benchmarking, you disable persistency completely.

```toml
[metadata]
schema_version = "2.0.0"

[logging]
threshold = "error"

[core]
listed = false
private = false
tracker_usage_statistics = false

[core.database]
driver = "sqlite3"
path = "./sqlite3.db"

[core.tracker_policy]
persistent_torrent_completed_stat = false
remove_peerless_torrents = false

[[udp_trackers]]
bind_address = "0.0.0.0:6969"
```

Even if you don't use the database at all, you need to add either the database configuration:

```toml
[core.database]
driver = "sqlite3"
path = "./sqlite3.db"
```

Or to create the default location for the database:

```
./storage/tracker/lib/database/
```

Otherwise, you get this error:

```
TORRUST_TRACKER_CONFIG_TOML_PATH=./share/default/config/tracker.udp.benchmarking.toml torrust-tracker
Loading extra configuration from file: `./share/default/config/tracker.udp.benchmarking.toml` ...
2024-08-07T14:37:42.754632Z ERROR r2d2: unable to open database file: ./storage/tracker/lib/database/sqlite3.db
```

Besides, the database is created and tables too.

I think the app should not:

- Require any DB configuration or storage folder.
- Or create any DB file or tables.

If none of the features requiring persistency is enabled.

cc @da2ce7

**Comment by josecelano** [comment]
I think the problem is we create the tables when we instantiate the DB driver:

```rust
pub fn build(driver: &Driver, db_path: &str) -> Result<Box<dyn Database>, Error> {
    let database = match driver {
        Driver::Sqlite3 => Builder::<Sqlite>::build(db_path),
        Driver::MySQL => Builder::<Mysql>::build(db_path),
    }?;

    database.create_database_tables().expect("Could not create database tables.");

    Ok(database)
}
```

I think we should move that logic out of the driver constructor and use an internal flag to call `create_database_tables` only once (the first time the driver is used).

**Comment by josecelano** [comment]
Phase 3 is implemented in PR #2099: https://github.com/torrust/torrust-tracker/pull/2099

This PR deliberately does not close this issue. It adds the v3 `Option<Database>` representation, optional composition seam, temporary v2 compatibility bridge, reusable bootstrap persistence-requirement matrix, and ADR while preserving the current persistence-initializing runtime.

Remaining delivery is staged:

1. Issue #1980 migrates active runtime consumers to v3 while retaining the explicit temporary database bridge.
2. The persistence-free runtime activation follow-up then replaces that bridge with the actual v3 `core.database` value, invokes the requirement validation, implements the no-driver/no-migration branch, updates the container entrypoint, and records M1–M6.
3. REST API persistence-aware public behavior remains deferred to API issue #144.

The activation follow-up has intentionally not been created as a GitHub issue yet. Its approved planning draft is `docs/issues/open/999-1978-optional-database-configuration/persistence-free-runtime-activation-draft.md`; it must be refined against merged #999 and #1980 implementation evidence before opening it.

**Comment by josecelano** [comment]
Closing this issue now that the staged delivery described above is complete.

- #2099 introduced the v3 `Option<Database>` configuration representation, composition seam, validation matrix, and temporary compatibility bridge.
- #1980 activated the v3 configuration at runtime while retaining that bridge.
- #2107, merged by PR #2112, replaced the bridge with the actual v3 `core.database` value and delivered the persistence-free runtime: no driver, migration, database file, or database connection when no configured capability requires persistence. It also completed the validation matrix, configured SQLite/MySQL/PostgreSQL lifecycle coverage, REST disabled-capability `409 Conflict` responses, container startup support, transition protection, and M1-M6 evidence.

The original outcome is therefore complete: a public tracker with persistence-backed capabilities disabled no longer requires database configuration or storage and does not create database artifacts. The remaining API provenance work is separately tracked under #144 and is not a blocker for this issue.

Closes #999 as completed.
